server/src/main/java/io/deephaven/server/console/completer/PythonAutoCompleteObserver.java

```diff
@@ -0,0 +1,234 @@
+                final VersionedTextDocumentIdentifier text = request.getTextDocument();
+
+                PyObject completer = (PyObject) scriptSession.get().getVariable("jedi_settings");
+                String uri = text.getUri();
+                int version = text.getVersion();
+                String document = completer.callMethod("get_doc", text.getUri()).getStringValue();
+
+                final List<ChangeDocumentRequest.TextDocumentContentChangeEvent> changes = request.getContentChangesList();
+                document = CompletionParser.updateDocumentChanges(uri, version, document, changes);
+                if (document == null) {
+                    return;
+                }
+
+                completer.callMethod("update_doc", document, uri, version);
+                break;
+            }
+            case GET_COMPLETION_ITEMS: {
+                GetCompletionItemsRequest request = value.getGetCompletionItems();
+                SessionState.ExportObject<ScriptSession> exportedConsole =
+                        session.getExport(request.getConsoleId(), "consoleId");
+                session.nonExport()
+                        .require(exportedConsole)
+                        .onError(responseObserver)
+                        .submit(() -> {
+                            getCompletionItems(request, exportedConsole, responseObserver);
+                        });
+                break;
+            }
+            case CLOSE_DOCUMENT: {
+                CloseDocumentRequest request = value.getCloseDocument();
+                PyObject completer = (PyObject) scriptSession.get().getVariable("jedi_settings");
+                completer.callMethod("close_doc", request.getTextDocument().getUri());
+                break;
+            }
+            case REQUEST_NOT_SET: {
+                throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT,
+                        "Autocomplete command missing request");
+            }
+        }
+    }
+
+    private void getCompletionItems(GetCompletionItemsRequest request,
+                                    SessionState.ExportObject<ScriptSession> exportedConsole,
+                                    StreamObserver<AutoCompleteResponse> responseObserver) {
+        final ScriptSession scriptSession = exportedConsole.get();
+        try (final SafeCloseable ignored = scriptSession.getExecutionContext().open()) {
+
+            PyObject completer = (PyObject) scriptSession.getVariable("jedi_settings");
+            boolean canJedi = completer.callMethod("is_enabled").getBooleanValue();
+            if (!canJedi) {
+                log.trace().append("Ignoring completion request because jedi is disabled").endl();
+                // send back an empty response. We may want to include a "don't ask again" flag in the response...
+                safelyExecuteLocked(responseObserver,
+                        () -> responseObserver.onNext(AutoCompleteResponse.newBuilder().build()));
```

Comment:
need to have the completed items response in here, with success=false, and requestId for the client to match the failure to

Reply:
fixed